### PR TITLE
Stinger Type Interface

### PR DESCRIPTION
**Pull request by rohitvarkey**
- [x] Julia type mapping to directly decode the Stinger C object.
- [x] Unexported function to provide the user with the `StingerGraph` instance with C values at that point of time.
- [x] Enumerations of the field types. This gives us the `StingerFields` type to use for functions to dispatch for these enums. It also allows us to figure out the index and calculate the offset easily.
- [x] `getindex` to read values of the C object in a lazy manner.
- [x] `setindex!` to write values to the C object.
- [x] Benchmarks for `getindex` and `setindex`
- [x] Insert remove edges benchmark
- [x] `outdegree` and `successors` implementations
- [x] BFS Implementation

I couldn't use `getfield` or `setfield!` and override the `a.b` syntax due to https://github.com/JuliaLang/julia/issues/1974.

Using `getindex` and `setindex!` along with the defined enumerations allows for usage like

``` julia
s = Stinger()
s[max_nv] = 1
s[max_nv] #Outputs 1
```

cc: @jpfairbanks

### CODE REVIEW

File: src/stinger_graph.jl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    vtype_names_start::Int64
+
+    ETA_start::Int64
+    ebpool_start::Int64
+    size_t::UInt
+end
+
+const stingergraphfields = fieldnames(StingerGraph)
+
+function createfieldenums()
```
Comment (jpfairbanks) [inline src/stinger_graph.jl]: This could use a descriptive comment for the next maintainer.

File: src/fields.jl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import Base: getindex, setindex!
+
+function getindex(x::Stinger, field::StingerFields)
+    idx = Int(field)
+    basepointer = convert(Ptr{fieldtype(StingerGraph, idx)}, x.handle)
```
Comment (jpfairbanks) [inline src/fields.jl]: You can remove this type introspection by specializing for the float case and int case separately. Although some benchmarks would prove if it is worth it.

File: src/stinger_graph.jl [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+const stingergraphfields = fieldnames(StingerGraph)
+
+"""
+Generates an Enumeration of all fields in `StingerGraph`.
+The generated fields are of type `StingerFields`.
+They are also exported so that the user can use it to access fields of `Stinger`
+objects using `getindex` and `setindex!`.
+"""
+function createfieldenums()
```
Comment (jpfairbanks) [inline src/stinger_graph.jl]: Can this be made a macro to avoid the eval?

**Review by jpfairbanks — COMMENTED** [review]
See the review notes.

File: src/stinger_graph.jl (hunk 3 above)
Comment (rohitvarkey) [reply src/stinger_graph.jl]: When I tried this as a macro, the exports where not coming in the global scope. I wrote the exports by hand which solved it for now.

**Comment by rohitvarkey** [comment]
The benchmarks for the type stable `getfield` vs the type instable `getfield` are shown here:

``` julia
julia> stableget = read(stable, "gettrial")
BenchmarkTools.Trial:
  samples:          10000000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     38.00 ns (0.00% GC)
  median time:      41.00 ns (0.00% GC)
  mean time:        43.95 ns (0.00% GC)
  maximum time:     219.11 μs (0.00% GC)

julia> unstableget = read(unstable, "gettrial")
BenchmarkTools.Trial:
  samples:          10000000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  32.00 bytes
  allocs estimate:  2
  minimum time:     366.00 ns (0.00% GC)
  median time:      398.00 ns (0.00% GC)
  mean time:        433.29 ns (1.10% GC)
  maximum time:     14.44 ms (99.79% GC)
```

I think this clearly shows that we should stick with the type stable version with the special functions for `batch_time` and `update_time`.

**Comment by rohitvarkey** [comment]
The insert remove benchmark here gives results of

Insert benchmark (attempting to insert 100000), terminates in 10 seconds.

``` julia
BenchmarkTools.Trial:
  samples:          84062
  evals/sample:     182
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  16.00 bytes
  allocs estimate:  1
  minimum time:     582.00 ns (0.00% GC)
  median time:      586.00 ns (0.00% GC)
  mean time:        650.31 ns (0.12% GC)
  maximum time:     21.48 μs (0.00% GC)
```

Remove benchmark (25000)

``` julia
BenchmarkTools.Trial:
  samples:          25000
  evals/sample:     178
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  16.00 bytes
  allocs estimate:  1
  minimum time:     599.00 ns (0.00% GC)
  median time:      604.00 ns (0.00% GC)
  mean time:        658.24 ns (0.14% GC)
  maximum time:     10.11 μs (91.55% GC)
```

**Comment by rohitvarkey** [comment]
The `outdegree` and `successors` implementation depends on https://github.com/stingergraph/stinger/pull/204 being merged.

**Pull request merged**